File: Robot/CommandDistributor.swift

import Foundation

extension InputStream {
	func read() -> UInt8? {
		var buffer = [UInt8]()
		guard self.read(&buffer, maxLength: 1) == 1 else {
			return nil
		}
		return buffer[0]
	}
}

/**
* Distributes Commands from an InputStream to CommandInterpreters
* which can be added for logging with "self.addCommandInterpreter(...)"
*/
/// Is a pseudo thread
public class CommandDistributor {
	
	
	private var inputStream: InputStream;
	private var commandInterpreters = Array<CommandInterpreter>();
	
	private var shouldStop = false;
	
	/**
	* @param inputStream from which commands get interpreted
	*/
	public init(inputStream: InputStream) {
		self.inputStream = inputStream;
		
		// start thread
		// FIXME make it async
		//self.start();
	}
	
	/**
	* notifies all command interpreters that they won't receive commands anymore
	* and removes them from the list of command interpreters
	*/
	// sync
	public func removeAllCommandInterpreters() {
		for commandInterpreter in commandInterpreters {
			commandInterpreter.willFinishInterpretingCommands();
		}
		commandInterpreters.removeAll();
	}
	
	/**
	* calls removeAllCommandInterpretes and stops reading from inputStream
	*/
	// sync
	public func stopThread() {
		removeAllCommandInterpreters();
		shouldStop = true;
	}
	
	/**
	* adds commandInterpeter to list of command interpreters
	* @param commandInterpreter which gets added
	*/
	// sync
	public func addCommandInterpreter(_ commandInterpreter: CommandInterpreter) {
		commandInterpreters.append(commandInterpreter);
	}
	
	
	public func run() {
		
		var commandString = "";
		
		// !!! can throw IOException
		while let c = inputStream.read() {
			
			if (shouldStop) {
				// stop thread
				return;
			}
			
			
			commandString.append(String(UnicodeScalar(c)));
			
			// sort out commandInterpreters which don't read commands
			var i = 0
			while i < commandInterpreters.count {
				if (commandInterpreters[i].finishedInterpretingCommands()) {
					commandInterpreters.remove(at: i);
					i -= 1
				}
				i += 1
			}
			
			if (Command.stringEndsWithEndDelimiter(commandString)) {
				
				var command: Command
				do {
					// !!! can throw IllegalArgumentException
					command = try Command.fromString(commandString)
				} catch {
					// continue reading from input Stream and reset command String
					commandString = "";
					continue;
				}
				
				// pass command to all commandInterpreters
				for commandInterpreter in commandInterpreters {
					commandInterpreter.interpretCommand(command);
				}
				
				// reset commandString
				commandString = "";
			}
			
		}
		
		
		commandInterpreters.forEach{
			$0.willFinishInterpretingCommands();
		}
		
		
		commandInterpreters.removeAll()
	}
	
}

File: Robot/Java/Helper/CommandDistributor.java
package Robot.Java.Helper;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Distributes Commands from an InputStream to CommandInterpreters
 * which can be added for logging with "this.addCommandInterpreter(...)"
 */
public class CommandDistributor extends Thread {

	private InputStream inputStream;
	private ArrayList<CommandInterpreter> commandInterpreters = new ArrayList<CommandInterpreter>();
	
	private boolean shouldStop = false;

	/**
	 * @param inputStream from which commands get interpreted
	 */
	public CommandDistributor(InputStream inputStream) {
		this.inputStream = inputStream;
		// start thread
		this.start();
	}
	
	/**
	 * notifies all command interpreters that they won't receive commands anymore
	 * and removes them from the list of command interpreters
	 */
	public synchronized void removeAllCommandInterpreters() {
		for (CommandInterpreter commandInterpreter : commandInterpreters) {
			commandInterpreter.willFinishInterpretingCommands();
		}
		commandInterpreters.clear();
	}
	
	/**
	 * calls removeAllCommandInterpretes and stops reading from inputStream
	 */
	public synchronized void stopThread() {
		removeAllCommandInterpreters();
		shouldStop = true;
	}
	
	/**
	 * adds commandInterpeter to list of command interpreters
	 * @param commandInterpreter which gets added
	 */
	public synchronized void addCommandInterpreter(CommandInterpreter commandInterpreter) {
		commandInterpreters.add(commandInterpreter);
	}

	@Override
	public void run() {
		try {

			String commandString = "";
			int c;

			// !!! can throw IOException
			while ((c = inputStream.read()) != -1) {
				
				if (shouldStop) {
					// stop thread
					return;
				}
				
				commandString += (char) c;

				// sort out commandInterpreters which don't read commands
				for (int i = 0; i < commandInterpreters.size(); i++) {
					if (commandInterpreters.get(i).finishedInterpretingCommands()) {
						commandInterpreters.remove(i);
						i--;
					}
				}
				
				if (Command.stringEndsWithEndDelimiter(commandString)) {
					
					Command command;
					try {
						// !!! can throw IllegalArgumentException
						command = Command.fromString(commandString);
					} catch (IllegalArgumentException e) {
						e.printStackTrace();
						// continue reading from input Stream and reset command String
						commandString = "";
						continue;
					}
					
					// FIXME
					/*if (!command.getCommandString().equals(Protocol.Heartbeat.heartbeat)) {
						System.out.println("Created command: " + command.convertToString());
					}*/
					
					// pass command to all commandInterpreters
					for (int i = 0; i < commandInterpreters.size(); i++) {
						commandInterpreters.get(i).interpretCommand(command);
					}
					
					// reset commandString
					commandString = "";
				}

			}
		} catch (IOException e) {
			for (CommandInterpreter commandInterpreter : commandInterpreters){
				commandInterpreter.willFinishInterpretingCommands();
			}
		}
		
		commandInterpreters.clear();
	}

}
